core/arch/arm/plat-imx/sm_platform_handler.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2017 NXP
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <io.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/tz_ssvce_pl310.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <sm/sm.h>
+#include <sm/optee_smc.h>
+
+#include "smc_sip.h"
+
+#ifdef CFG_PL310
+static void handle_sip_pl310(struct thread_smc_args *smc_args)
+{
+	smc_args->a0 = OPTEE_SMC_RETURN_OK;
+	DMSG("SIP L2C310 val=%x addr=%x\n", smc_args->a1,
+	     smc_args->a2);
+	/*
+	 * Disable L2 is only needed in low power case,
+	 * the psci asm code will take care of disable.
+	 * So we just flush data here.
+	 */
+	if (smc_args->a1 == 0 && smc_args->a2 == PL310_CTRL) {
+		dcache_op_all(DCACHE_OP_CLEAN);
```

Comment:
What happens if normal world skips this before suspend? Could it lead to corruption in secure world?

Reply:
Actually no, I think. This is to keep safe, so add it here, also to support outer_disable/resume. The final psci asm code will sync L2 again.

Reply:
Sorry for being a bit stubborn, but how does this make things more safe if the suspend code will do it too? I'd like to avoid eventual grey zones.

Reply:
@jenswi-linaro Previously, I do not want to introduce the SIP PL310, and we let Linux unlock pl310 ways when initializating. But ARM Linux maintainer rejected this. So I start to add SIP PL310 support and use l2c_write_sec in linux side.  So we do not touch pl310 enable/configuration in OP-TEE directly unless linux request it.
Now talking suspend/resume in Linux side,  the flow now is
```
outer_disable();
cpu_suspend(xxxxx);
outer_resume
```
when Linux kernel runs into suspend by "echo mem > /sys/power/state",  the upper flow will start.
outer_disable will disable L2 with flush L2, outer_resume will enable L2 with invalidation ways. On i.MX6, we did not power off L2 memory when suspend, we only power off L2 controller. Because outer_resume will invalidate L2 memory, I use `dcache_op_all(DCACHE_OP_CLEAN); arm_cl2_cleanbyway(pl310_base())` in L2 disable process to make sure data is synced to memory. Actually in the finaly psci asm code, I'll do another sync. I do not very trust the C code doing cache flush.

When normal world suspending, it will first freeze all the user process, it means all active OP-TEE normal world process will need to  finish it's work and runs into freeze, if the optee threads refuse to freeze, suspend be aborted after a while, about 60s I think. 

Now back to your question, if normal world skipping L2 disable.  If normal world wants to be in suspended state, it must call psci flow into optee to suspend. In psci asm code, we will flush L1/L2 again to make sure data is flushed to memory. Flushing cache in secure world will flush non-secure/secure cache all. So it will make sure linux/optee data in cache all flushed to memory. Even Linux does not flush cache, the psci asm code will also keep it flushed.

But my case is I need to support Linux running in secure and non-secure world both. So some piece code is duplicated.

Hope this is clear.

Reply:
OK, thanks I understand now.

Reply:
As this only does not disable the outer cache instead of disabling it, won't the cache be refilled with non-secure data once back in REE (in the `cpu_suspend(xxx)` sequence) whereas REE would expect it to be disabled?